### PR TITLE
Validate the refs exists

### PR DESCRIPTION
**Pull request by ribeiromiranda**
Add validator with context to validate the refs (functions, events, retries, etc) exists

### CODE REVIEW

**Comment by ribeiromiranda** [comment]
@ricardozanini To solve the unit test that PR (involve the default values of `Function.Operation`), I want to wait for PR #159.

**Comment by ricardozanini** [comment]
#159 should be merged soon.

**Comment by ricardozanini** [comment]
@ribeiromiranda any updates on this one?

**Comment by ribeiromiranda** [comment]
@ricardozanini Yes, but I didn't push it because there's a lot of unit testing to fix.

File: model/workflow_validator.go [hunk 1]
```diff
@@ -224,3 +220,77 @@ func validationNotExclusiveParamters(values []bool) bool {
...
+
+func (e WorkflowErrors) Error() string {
+	errors := []string{}
+	for _, err := range []error(e) {
+		errors = append(errors, err.Error())
+	}
+	return strings.Join(errors, "\n")
+}
+
+func WorkflowError(err error) error {
```
Comment (ribeiromiranda) [inline model/workflow_validator.go:234]: Hey guys. Also start in that PR the issue https://github.com/serverlessworkflow/sdk-go/issues/125. Can you validate these error messages are ok?

File: model/workflow_validator.go [hunk 2]
```diff
@@ -15,83 +15,282 @@
...
 	}
+}
+
+func (c *mapValues[T]) contain(name string) bool {
+	_, ok := c.ValuesMap[name]
+	return ok
+}
+
+func init() {
+	// TODO: create states graph to complex check
```
Comment (ricardozanini) [inline model/workflow_validator.go:131]: Can we have a follow-up issue to have it documented?

File: model/workflow_validator.go [hunk 3]
```diff
@@ -15,83 +15,282 @@
...
+			if part != "Workflow" && part != "BaseWorkflow" && part != "BaseState" {
+				part := strings.ToLower(part[:1]) + part[1:]
+				newNamespaceList = append(newNamespaceList, part)
+			}
+		}
+		namespace := strings.Join(newNamespaceList, ".")
+
+		switch err.Tag() {
+		case "exists":
+			workflowErrors = append(workflowErrors, fmt.Errorf("%s don't exists %q", namespace, err.Value()))
```
Comment (ricardozanini) [inline model/workflow_validator.go:271]:
```suggestion
			workflowErrors = append(workflowErrors, fmt.Errorf("%s don't exist %q", namespace, err.Value()))
```

File: model/workflow_validator.go [hunk 4]
```diff
@@ -15,83 +15,282 @@
...
+		case TagCompensatedby:
+			workflowErrors = append(workflowErrors, fmt.Errorf("%s compensatedBy don't exists %q", namespace, err.Value()))
```
Comment (ricardozanini) [inline model/workflow_validator.go:273]:
```suggestion
			workflowErrors = append(workflowErrors, fmt.Errorf("%s compensatedBy don't exist %q", namespace, err.Value()))
```

File: model/workflow_validator.go [hunk 5]
```diff
@@ -15,83 +15,282 @@
...
+		case "unique":
+			if err.Param() == "" {
+				workflowErrors = append(workflowErrors, fmt.Errorf("%s has duplicate value", namespace))
+			} else {
+				workflowErrors = append(workflowErrors, fmt.Errorf("%s has duplicate %q", namespace, strings.ToLower(err.Param())))
+			}
+		case "required_without":
+			if err.Param() == "ID" {
+				workflowErrors = append(workflowErrors, errors.New("id required when not defined \"key\""))
```
Comment (ricardozanini) [inline model/workflow_validator.go:282]:
```suggestion
				workflowErrors = append(workflowErrors, errors.New("id is required when \"key\" is not defined "))
```

File: model/workflow_validator.go [hunk 6]
```diff
@@ -15,83 +15,282 @@
...
+			} else if err.Param() == "Key" {
+				workflowErrors = append(workflowErrors, errors.New("key required when not defined \"id\""))
```
Comment (ricardozanini) [inline model/workflow_validator.go:284]:
```suggestion
				workflowErrors = append(workflowErrors, errors.New("key is required when \"id \" is not defined"))
```

**Review by ricardozanini — COMMENTED** [review]
@ribeiromiranda, this is awesome work. Many many thanks for this. I imagine you're still working on it, right?

Apart from that, can you join the CNCF Slack, #serverless-workflow-sdk channel? I need to talk to you. Just send me a PM. :)

File: model/states_validator_test.go [hunk 7]
```diff
@@ -76,60 +79,52 @@ var switchStateTransitionDefault = State{
...
-	type testCase struct {
-		name     string
-		instance State
-		err      string
-	}
+	// type testCase struct {
+	// 	name     string
+	// 	instance State
+	// 	err      string
+	// }
```
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply model/states_validator_test.go]: Yes. I remove.

File: util/unmarshal.go [hunk 8]
```diff
@@ -188,9 +189,10 @@ func getBytesFromHttp(url string) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply util/unmarshal.go]: @sonatype-lift ignore

File: util/unmarshal.go [hunk 9]
```diff
@@ -174,7 +175,7 @@ func getBytesFromHttp(url string) ([]byte, error) {
 		return nil, err
 	}
 
-	resp, err := httpClient.Do(req)
+	resp, err := HttpClient.Do(req)
```
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply util/unmarshal.go:178]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply util/unmarshal.go:178]: @sonatype-lift ignore

File: model/workflow_validator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply model/workflow_validator.go]: I agree.

**Comment by ricardozanini** [comment]
@ribeiromiranda are you still working on this, right? Please let me know when ready so we can have a final look + merge.

**Comment by ribeiromiranda** [comment]
> @ribeiromiranda are you still working on this, right? Please let me know when ready so we can have a final look + merge.

@ricardozanini No more changes.

**Comment by ricardozanini** [comment]
> > @ribeiromiranda are you still working on this, right? Please let me know when ready so we can have a final look + merge.
> 
> @ricardozanini No more changes.

Can you take a look at the CI errors, then? So we can merge after it. cc @spolti

File: model/retry.go [hunk 10]
```diff
@@ -41,3 +42,15 @@ type Retry struct {
...
+type retryUnmarshal Retry
+
+// UnmarshalJSON implements json.Unmarshaler
+func (r *Retry) UnmarshalJSON(data []byte) error {
+	r.ApplyDefault()
+	return util.UnmarshalObject("retry", data, (*retryUnmarshal)(r))
+}
+
+func (r *Retry) ApplyDefault() {
+	r.MaxAttempts = intstr.FromInt(1)
```
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply model/retry.go:55]: @sonatype-lift ignore

**Comment by ribeiromiranda** [comment]
@ricardozanini @spolti CI errors fixed.

File: validator/workflow.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+		// fmt.Println("StructNamespace", err.StructNamespace())
+		// fmt.Println("StructField", err.StructField())
+		// fmt.Println("Tag", err.Tag())
+		// // fmt.Println(err.ActualTag())
+		// // fmt.Println(err.Kind())
+		// // fmt.Println(err.Type())
+		// fmt.Println("value", err.Value())
+		// fmt.Println("param", err.Param())
+		// fmt.Println()
+
```
[the remark this answers is not in the document]
Comment (ribeiromiranda) [reply validator/workflow.go]: Removed

File: model/workflow_validator.go [hunk 12]
```diff
@@ -15,83 +15,216 @@
...
+
+func (c *ValidatorContext) ExistError(name string) bool {
+	_, ok := c.Errors[name]
+	return ok
+}
+
+func NewValidatorContext(workflow *Workflow) context.Context {
+	for i := range workflow.States {
+		s := &workflow.States[i]
+		if s.BaseState.Transition != nil {
```
Comment (spolti) [inline model/workflow_validator.go:104]: isn't a switch better here?

Comment (spolti) [reply model/workflow_validator.go:104]: it is not. :)

**Review by spolti — APPROVED** [review]
Really nice :)

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
Waiting for #178

**Comment by ribeiromiranda** [comment]
> Waiting for #178

@ricardozanini  Made the rebase with the main branch.

**Pull request merged**